src/webgpu/shader/execution/expression/call/builtin/textureStore.spec.ts

```diff
@@ -120,3 +120,377 @@ Parameters:
+  return ${type}(id);
+}`;
+      break;
+    case '2d':
+      return `
+fn indexToCoord(id : u32) -> vec2<${type}> {
+  return vec2<${type}>(${type}(id % width), ${type}(id / width));
+}`;
+      break;
+    case '3d':
+      return `
+fn indexToCoord(id : u32) -> vec3<${type}> {
+  const half = numTexels / depth;
+  let half_id = id % half;
+  return vec3<${type}>(${type}(half_id % width), ${type}(half_id / width), ${type}(id / half));
+}`;
+      break;
+  }
+  return ``;
+}
+
+// Mutates 'coords' to produce an out-of-bounds value.
+// 1D workgroups are launched so 'gid.x' is the linear id.
+function outOfBoundsValue(dim: GPUTextureDimension, type: string): string {
+  switch (dim) {
+    case '1d': {
+      if (type === 'i32') {
+        return `if gid.x % 3 == 0 {
+          coords = -coords;
+        } else {
+          coords = coords + numTexels;
+        }`;
+      } else {
+        return `coords = coords + numTexels;`;
+      }
+      break;
+    }
+    case '2d': {
+      if (type === 'i32') {
+        return `if gid.x % 3 == 0 {
+          coords.x = -coords.x;
+        } else {
+          coords.y = coords.y + height;
+        }`;
+      } else {
+        return `if gid.x % 3 == 1 {
+          coords.x = coords.x + width;
+        } else {
+          coords.y = coords.y + height;
+        }`;
+      }
+      break;
+    }
+    case '3d': {
+      if (type === 'i32') {
+        return `if gid.x % 3 == 0 {
+          coords.x = -coords.x;
+        } else if gid.x % 5 == 0 {
+          coords.y = coords.y + height;
+        } else {
+          coords.z = coords.z + depth;
+        }`;
+      } else {
+        return `if gid.x % 3 == 1 {
+          coords.x = coords.x + width;
+        } else if gid.x % 5 == 1 {
+          coords.y = coords.y + height;
+        } else {
+          coords.z = 2 * depth;
+        }`;
+      }
+      break;
+    }
+  }
+  return ``;
+}
+
+function getMipTexels(numTexels: number, dim: GPUTextureDimension, mip: number): number {
+  let texels = numTexels;
+  if (mip === 0) {
+    return texels;
+  }
+  if (dim === '2d') {
+    texels /= 1 << mip;
+    texels /= 1 << mip;
+  } else if (dim === '3d') {
+    texels /= 1 << mip;
+    texels /= 1 << mip;
+    texels /= 1 << mip;
+  }
+  return texels;
+}
+
+const kDims = ['1d', '2d', '3d'] as const;
+
+g.test('out_of_bounds')
+  .desc('Test that textureStore on out-of-bounds coordinates have no effect')
+  .params(u =>
+    u
+      .combine('dim', kDims)
+      .combine('coords', ['i32', 'u32'] as const)
+      .combine('mipCount', [1, 2, 3] as const)
+      .combine('mip', [0, 1, 2] as const)
+      .filter(t => {
+        if (t.dim === '1d') {
+          return t.mipCount === 1 && t.mip === 0;
+        }
+        if (t.dim === '3d') {
+          return t.mipCount <= 2 && t.mip < t.mipCount;
+        }
+        return t.mip < t.mipCount;
+      })
+  )
+  .fn(t => {
+    const texel_format = 'r32uint';
+    const num_texels = 4096;
```

Comment:
This 4096 is chosen to be big enough to ensure x,y,and z dimensions are non-trivial.
Can you show the math in the worst case in comments?